### DISCUSSION TITLE
SQL database performance in Docker containers vs natively on the host

### DISCUSSION THREAD

**Discussion by Xitric** [testcontainers/testcontainers-dotnet#625] [General]
We have started to look into using Docker containers in general for supporting all of our integration and benchmark testing projects due to the ease of use for us as developers (not having to configure databases locally and manage connection strings, worrying about ports, etc.). This is really cool, but we have unfortunately also experienced a degradation in performance of all testing projects that involve MsSql databases hosted in Docker containers.

This is not an issue with testcontainers, since we experience the same degradation when simply using Docker compose to prepare our databases.

To give an example, we have certain database-heavy integration tests that went from executing in 1.5 seconds when the database was installed natively on the host, to spending 5-6 seconds when the database was in a Docker container.

We also have some benchmark tests that are extremely database-heavy, and here we see a performance degradation of a factor 10.

Here is some additional information for context:

- All developers use Windows 10 together with Docker Desktop
- Our docker containers do not hit the resource limits during the tests
- We are not doing anything fancy with our database containers:
    - Image: `mcr.microsoft.com/mssql/server:2019-CU11-ubuntu-20.04`
    - A port binding
    - Username and password via environment variables

We suspect there is some form of overhead, perhaps related to the Windows file system of the host, which affects the performance of our tests when using a Docker container to host our databases. Certain things we have tried so far:

- Allocating 20 GB of memory and 14 CPU cores to the database containers alone, with little to no measurable impact on performance
- Setting the `WithPrivileged` flag on the Docker container, with little to no measurable impact on performance
- Reduce consistency of all volumes mounted to the SQL container to either delegated or cached, with a small improvement to performance, but not enough to make up for the degradation

But we would really like to hear if anyone else in the community has experienced any performance issues when running any type of SQL server in a Docker container, and what, if anything, they have done to improve upon this!

**Comment by HofmeisterAn** [comment]
What setup do you use? A single shared instance or multiple parallel instances? How do you access the database? With the native client? What is the MSSQL max server memory? @kiview @eddumelendez Do you have ideas or maybe made experiences that could reduce the container performance?

**Reply by Xitric** [reply]
Let me ty to answer those questions with as much detail as I can.

**A single shared instance or multiple parallel instances?**

We have tried both in fact. A single shared instance of MsSql using Docker compose, and 5 individual MsSql instances running in parallel. The 5 MsSql containers then share the 20 GB of memory and the 14 cores. I suspect that is way over the top, but we tried to push it to see if it would have an effect. This is when we run the complete integration test suite.

However, we have also tried to execute individual tests one at a time, sometimes in a loop, to debug performance issues. And the result is the same. Executing a single test once with a natively installed database versus one in a Docker container results in a degradation of at least a factor 3 for our database-heavy integration tests. There is no particular load on the database during such a single test.

**How do you access the database?**

We access the database from our integration and benchmark tests written in .NET and using Dapper+System.Data.SqlClient for the integration tests (for now) and EF Core+Microsoft.Data.SqlClient for the benchmark tests.

For all tests, I believe we use a single shared database connection for the lifetime of a test assembly with a bit of parallellization inside the test assembly. With the benchmark tests, we create a new database context for every iteration of a benchmark. But the database connection is managed by us and fed to EF Core.

If you need more precise insight into the design and setup of our tests just let me know, and I will involve some of my colleagues in this discussion.

**What is the MSSQL max server memory?**

I don't know, so I would need to investigate this further. But given that even individual tests exhibit the same degradation we observe when we run multiple tests either sequentially or a few in parallel, I don't believe it is a memory issue.

**Reply by HofmeisterAn** [reply]
Difficult to help without being able to look and debug into the code, sharing ideas and maybe best practices might still help. Any I/O bottleneck (disk)? Do you run them with xUnit? Do the tests run in the same collection (that would run them sequential and not parallel)?

First, I would take a closer look into the default MSSQL container settings and check:

- https://learn.microsoft.com/en-us/sql/linux/sql-server-linux-performance-best-practices?view=sql-server-2017
- https://learn.microsoft.com/en-us/sql/linux/sql-server-linux-configure-environment-variables?view=sql-server-2017
- https://learn.microsoft.com/en-us/sql/linux/sql-server-linux-docker-container-deployment?view=sql-server-2017&pivots=cs1-bash#production

Then I would take a look into parallelization, 3x sounds too much.

**Reply by Xitric** [reply]
You are right, I will try to extract a reproducible baseline from our testing harness so that we have something to base the investigation around.

In the meantime I will also read the resources you  shared to see if something might be of help for us. Thank you for helping out so far!

**Reply by HofmeisterAn** [reply]
One thing you might check as well is, if you attach an output consumer. Reading the stream  all the time may reduce performance too.

**Reply by jgeorgie** [reply]
@HofmeisterAn hello! 
We've created a small repo, where you can reproduce the issue. You'll just need to run the integration tests there. One of them is using SQL server local instance, another one is using testcontainers. 
Here is the link:
https://github.com/jgeorgie/SQLPerformanceDockerVSnative

**Reply by HofmeisterAn** [reply]
Haven't forgot it, just some other topics came up. I will need some more days until I can take a look at it.

**Reply by Xitric** [reply]
No worries, we are simply grateful for any help we can get on this topic. We are not blocked by this in any way yet.

**Reply by Xitric** [reply]
@HofmeisterAn Sorry for pinging you, but did you ever have time to look into this? We are still having trouble with the performance of our SQL based integration tests when running the SQL server in Docker containers.

**Reply by HofmeisterAn** [reply]
Yes, I started working on it weeks ago, but then got distracted 😬. Just recently, I disabled durability for PostgreSQL. According to this [article](https://shusson.info/post/non-durable-postgres-for-e2e-tests), it has reduced the test time by 20%. Do you know if there are similar optimizations for MSSQL?

**Reply by HofmeisterAn** [reply]
I spent some time looking at the example project and remembered that I had trouble getting the test running. Specifically, the code that prepares the performance test (database) did not work as expected - for example, `CreateTable(string)` did not create any table. Additionally, I was unsure if `StartTest(SqlConnection)` was running as intended. I had to make some changes to get it working, and I noticed that the example project could benefit from following more best practices and being cleaned up. Altogether, the example appears too complicated and disorganized (that is why I did not start further analysis and performance investigations).

In particular, I would recommend cleaning up the project and perhaps replacing `MsSqlTestcontainer` with a proper module implementation, then re-testing it. Out of curiosity, how long did your test run for? In my case, it took less than 300ms.

**Reply by Xitric** [reply]
I believe the tests in our project in general leave a lot to desire in terms of best practices. They are unfortunately also ridiculously complicated. I think the reason why the sample project appears chaotic is that we attempted to extract the parts of our test harness that leave a measurable impact on performance, without exposing the inner details of our software itself.

We can certainly try to use the built-in MSSQL test container now. We could not originally, when we started implementing test container support in our projects, but with some of the recent changes to that module I believe it now supports all of our needs.

I will also check to see if we can remove additional code from that sample project while still being able to reproduce the performance issues, and then give you some information on the execution times that we observe.

**Reply by HofmeisterAn** [reply]
> We can certainly try to use the built-in MSSQL test container now.

Either you can wait for the next release, or you can copy the module configuration. However, I don't believe that it will make a significant difference. The configurations are similar, and I don't expect any performance improvements. It will simply be a little cleaner.

Did I understand correctly that you are benchmarking against `ALTER TABLE ? NOCHECK CONSTRAINT ALL`? Can you share the results of your test? I haven't done a lot of database development in the past, so I'm not sure if that is a good performance indicator, although it may show significant differences.

Have you considered reaching out to Microsoft? It seems that you're not the only one experiencing this issue: https://github.com/microsoft/mssql-docker/issues/644. I believe that a proper test with the support of TC could help to identify the bottleneck, and it would be a good way to benchmark it.

**Comment by pfaustinom1** [comment]
Did you manage to solve this problem? We’re running into exactly the same issue: running a database migration against a Testcontainers-hosted SQL Server is much slower than running the same migration against a native instance on the host. Our setup is Windows 11 with Docker Desktop. Here’s exactly how we spin up SQL Server in Testcontainers:

```csharp
sqlServerContainer = new MsSqlBuilder()
  .WithImage("mcr.microsoft.com/mssql/server:2019-latest")
  .WithPassword(sqlServerPassword)
  .WithBindMount(localBackupDirectory, "/backup", AccessMode.ReadOnly)
  .WithCleanUp(true)
  .Build();
```
We point our migration code at the connection string provided by Testontainer and see a several-second delay compared to running the same migration against a natively installed SQL Server.
